src/Middleware/RateLimiting/src/RateLimitingMiddleware.cs

```diff
@@ -55,19 +79,126 @@ public async Task Invoke(HttpContext context)
             // OnRejected "wins" over DefaultRejectionStatusCode - we set DefaultRejectionStatusCode first,
             // then call OnRejected in case it wants to do any further modification of the status code.
             context.Response.StatusCode = _rejectionStatusCode;
-            await _onRejected(context, lease);
+
+            // If the global limiter rejected this request, use OnRejected from the options if it exists.
+            // Else the endpoint limiter rejected this request - use OnRejected from the endpoint policy if it exists,
+            // else OnRejected from the options if it exists.
+            if (leaseContext.Rejector.Equals(Rejector.Global))
+            {
+                if (_onRejected is not null)
+                {
+                    await _onRejected(new OnRejectedContext() { HttpContext = context, Lease = leaseContext.Lease }, context.RequestAborted);
```

Comment:
nit: would be cleaner to factor out the `OnRejectedContext` construction and move it above since it's instantiated the same way in all three of the branches around here.

Reply:
Could go even further and make a local variable to hold the onRejected function and then assign the variable in the conditions and at the end call the local onRejected variable in one place instead of three different places